### PR TITLE
Rework Ring-bearer implementation. Add GUI + gamelogs.

### LINKED ISSUE #10453: 'The Ring Tempts You' effects never triggers more than once
'The Ring' will only tempt you once no matter how many 'The Ring Tempts You' effects you play, so you will never be tempted 4 times to get a fully powered 'The Ring' emblem

Edit: It may just be the Nazgul that is bugged, specifically playing multiple of them

### LINKED ISSUE #10572: No indicator of which creature is the Ring bearer
Can we get a ring icon or something that indicates the Ring bearer? And / or have the ring emblem indicate which creature (with the UUID) is the Ring bearer? Currently I have to look at the tool tip and see if the creature is legendary or not. However, this doesn't work if all your creatures are legendary anyways.

### PR DESCRIPTION
**Pull request by Susucre**
Aside from the cog being placeholder (I can try to make a ring.svg, but do not hope for much),
this should improve the feedback on what happened on ring tempting a player (with logs, and info + icon on the ringbearer) 

In addition, there was a few wrong paths in the ring bearer chosing logic.
If there is at least one creature, there should always be a ring bearer chosen event.

I pondered adding an info on the ring emblem to tell that there is (or there is none) Ringbearer at the moment.
Did not feel critical enough.

ringtempting
ringbearing

close #10572 (once there is a proper svg.)
fix #10453 maybe ? The logic has been redone, and there is now game logs to be sure that something that should have happen did not.

### CODE REVIEW

**Comment by thefnitalian** [comment]
Use https://www.vectorstock.com/royalty-free-vector/diamond-ring-icon-vector-33131789

**Comment by Susucre** [comment]
Won't do, I think we specifically need svg, and probably of a set ratio.

**Comment by JayDi85** [comment]
[readme-icons.txt](https://github.com/magefree/mage/blob/master/Mage.Client/src/main/resources/card/icons/readme-icons.txt) contains recommendation and sources for the icon resources.

Good example:
https://icons.getbootstrap.com/icons/record-circle-fill/
shot_230708_234501



BTW icons engine support any sides and any order for the icons (left, center, right, combined order). Test tools contains test render dialog with icons support (must enable cards with icons in the code to see in it).

If ring status is a critical for the whole game then it can be put to the top as except. Also it can be with custom yellow/gold color. See commander icon for example.

**Comment by alexander-novo** [comment]
https://icons8.com/icon/41528/one-ring

This one has an attribution license. Just need to add a link in an about page

**Comment by Susucre** [comment]
Thanks @alexander-novo , really like that one.
Since I do want it to stand out, it needs to be on another panel to have its own color. I positioned it right for now, if someone has strong opinions about it, feel free to change it later on.
rightsideringbearer

As for the attribution license, they seem to be fine with open source projects, I did add it in the readme, and the source svg in a new dedicated folder. Is that enough?

**Comment by alexander-novo** [comment]
https://icons8.com/license

No, they specify that desktop applications must have a link in the about section

**Comment by Susucre** [comment]
Uhm I'll try to find how to modify it then

**Comment by JayDi85** [comment]
icons8 is weird source for the project and must be removed. Use sources with cc/mit license styles without things like "Add a link to icons8.com on all pages using our content. If that's most pages, a link in your footer is fine."

**Comment by alexander-novo** [comment]
I think it's fine to get rid of this as a source, but I do want to point out the desktop application section, rather than the website section.

**Comment by Susucre** [comment]
Going to use this one: https://github.com/FortAwesome/Font-Awesome/blob/6.x/js-packages/%40fortawesome/fontawesome-free/svgs/solid/ring.svg

**Comment by alexander-novo** [comment]
Oh that looks good too

**Comment by JayDi85** [comment]
> There are only so many sides to cards.

Yes, the main idea behind card panels -- icons from one panel can be auto-sorted and combined (if it catch too much items):
 - if you put one icon to the centred panel then you got "x x O x x"
 - if you put two icons to the centred panel then you got "x O O x x"
 - if you put three icons to the centred panel then you got "x O O O x"

BTW right side/corner of the card will be used for counters someday, see #8149

**Review by JayDi85 — CHANGES_REQUESTED** [review]

File: Mage/src/main/java/mage/players/PlayerImpl.java [hunk 1]
```diff
@@ -5155,44 +5158,71 @@ public Permanent getRingBearer(Game game) {
             return bearer;
         }
         ringBearerId = null;
+        bearer.addInfo(ringbearerInfoKey, null, game);
```
[the remark this answers is not in the document]
Comment (Susucre) [reply Mage/src/main/java/mage/players/PlayerImpl.java:5161]: There is two problems with that:
- The ring bearer status is lost when you lose the control of the permanent. I think that was the intent for clearing ringBearerId in that function.
- Ring Bearer status has nothing to do with layers. We could reimplement it as a tapped status by adding a boolean isRingBearer to all Permanents if that helps with the engine.

**Comment by Susucre** [comment]
We could use an icon color overwrite inside an icon panel. I will not bother with it in this PR however. The red used for the commander icon is very bright and the ring svg is a little too wide for that color.
ringred

For reference, the GOLD color added looks like that. Not that great, but less in your face.
ring

**Comment by alexander-novo** [comment]
I think it could use some combination of the SVG being scaled down a bit and the top row moving up a little bit as right now while the crown doesn't get drawn over the text, the ring does and both of them conflict with the mana cost.

File: Mage/src/main/java/mage/players/PlayerImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (JayDi85) [reply Mage/src/main/java/mage/players/PlayerImpl.java:5161]: > I think that was the intent for clearing ringBearerId in that function.

well, I see... There is much more complex problem with original implementation.

Restore code contains:
`this.ringBearerId = player.getRingBearerId();`

But `getRingBearerId` writing some code. Copy and restore constructors must be simple, without any additional/writing logic in the middle of the restore procedure. So bearer modification code must be removed at all.

For new implementation:
* 701.52a Certain spells and abilities have the text “the Ring tempts you.” Each time the Ring tempts you, choose a creature you control. That creature becomes your Ring-bearer **until another creature becomes your Ring-bearer or another player gains control of it**.
* 701.52b Ring-bearer is a designation **a permanent can have**. Being a Ring-bearer is not a copiable value.
* So permanent must have a flag, not a player;
* Flag must not be a copyable for copy effects;
* `getRingBearerId` must be dynamic with Game param to find an actual id;
* bearer set in two places: on new bearer choose, on `checkControlChanged` (it must call additional method like `removeUncontrolledBearer`).

[the remark this answers is not in the document]
Comment (Susucre) [reply Mage/src/main/java/mage/players/PlayerImpl.java]: Alright, that ended up less complex that I thought.
Please let me know if anything is wrong in the new implementation.
The isRingbearer logic is now at the charge of the PermanentImpl class.

**Comment by Susucre** [comment]
Added another game log message on control loss of the ring bearer:
lostcontrolringbearer

This has become more of a general rework of the mechanic that just GUI improvment.
I'll let the fine tuning of the ring icon to another PR, I am not great at the whole svg tweaking.

File: Mage/src/main/java/mage/players/PlayerImpl.java [hunk 2]
```diff
@@ -5140,59 +5135,85 @@ public FilterMana getPhyrexianColors() {
...
+
+        if(currentBearerId != null && currentBearerId == newBearerId) {
+            // Oracle Ruling for Call of the Ring
+            //
+            // If the creature you choose as your Ring-bearer was already your Ring-bearer,
+            // that still counts as choosing that creature as your Ring-bearer for the purpose
+            // f abilities that trigger "whenever you choose a creature as your Ring-bearer"
+            // or abilities that care about which creature was chosen as your Ring-bearer.
+            // (2023-06-16)
+            game.fireEvent(GameEvent.getEvent(GameEvent.EventType.RING_BEARER_CHOSEN, currentBearer.getId(), null, getId()));
```
Comment (JayDi85) [inline Mage/src/main/java/mage/players/PlayerImpl.java:5206]: miss null check for currentBearer

File: Mage/src/main/java/mage/players/PlayerImpl.java [hunk 3]
```diff
@@ -5140,59 +5135,85 @@ public FilterMana getPhyrexianColors() {
...
+            game.informPlayers(getLogName() + " did not choose a new Ring-bearer. " +
+                "It is still " + currentBearer.getLogName() + ".");
+        } else {
+            Permanent ringBearer = game.getPermanent(newBearerId);
             game.fireEvent(GameEvent.getEvent(GameEvent.EventType.RING_BEARER_CHOSEN, newBearerId, null, getId()));
-            this.ringBearerId = newBearerId;
+            game.informPlayers(getLogName() + " has chosen " + ringBearer.getLogName() + " as Ring-bearer.");
```
Comment (JayDi85) [inline Mage/src/main/java/mage/players/PlayerImpl.java:5212]: miss null check for ringBearer

File: Mage/src/main/java/mage/players/PlayerImpl.java [hunk 4]
```diff
@@ -5140,59 +5135,85 @@ public FilterMana getPhyrexianColors() {
...
+
+            // The setRingBearer method is taking care of removing
+            // the status from the current ring bearer, if existing.
+            ringBearer.setRingBearer(game, true);
         }
```
Comment (JayDi85) [inline Mage/src/main/java/mage/players/PlayerImpl.java:5217]: `RING_BEARER_CHOSEN` event must be called after `setRingBearer`, so triggers will see actual state of the battlefield and ring bearers.

**Review by JayDi85 — APPROVED** [review]
Some little fixes needs, but all other are fine now.

**Review by xenohedron — APPROVED** [review]

**Pull request merged**